src/connecting_stream.rs

```diff
@@ -1,69 +1,162 @@
-use std::{io, net::ToSocketAddrs};
+use std::io;
 
 use futures::{future::FutureResult, SelectOk};
 use tokio::net::{tcp::ConnectFuture, TcpStream};
 use tokio::prelude::*;
+use url::Url;
 
-enum State {
+use native_tls::TlsConnector;
+use tokio_tls::TlsStream;
+
+use crate::{
+    errors::ConnectionError,
+    io::Stream as InnerStream,
+};
+
+type ConnectingFuture<T> = Box<dyn Future<Item = T, Error = ConnectionError> + Send>;
+
+impl From<io::Error> for ConnectionError {
+    fn from(e: io::Error) -> Self {
+        Self::IoError(e)
+    }
+}
+
+impl From<native_tls::Error> for ConnectionError {
+    fn from(e: native_tls::Error) -> Self {
+        Self::TlsError(e)
+    }
+}
+
+enum TcpState {
     Wait(SelectOk<ConnectFuture>),
-    Fail(FutureResult<TcpStream, io::Error>),
+    Fail(FutureResult<TcpStream, ConnectionError>),
+}
+
+enum TlsState {
+    Wait(ConnectingFuture<TlsStream<TcpStream>>),
+    Fail(FutureResult<TlsStream<TcpStream>, ConnectionError>),
+}
+
+enum State {
+    Tcp(TcpState),
+    Tls(TlsState),
 }
 
 impl State {
-    fn poll(&mut self) -> Poll<TcpStream, io::Error> {
+    fn poll(&mut self) -> Poll<InnerStream, ConnectionError> {
         match self {
-            State::Wait(ref mut inner) => match inner.poll() {
-                Ok(Async::Ready((tcp, _))) => Ok(Async::Ready(tcp)),
-                Ok(Async::NotReady) => Ok(Async::NotReady),
-                Err(err) => Err(err),
+            Self::Tcp(state) => {
+                match state {
+                    TcpState::Wait(ref mut inner) => match inner.poll() {
+                        Ok(Async::Ready((tcp, _))) => Ok(Async::Ready(tcp.into())),
+                        Ok(Async::NotReady) => Ok(Async::NotReady),
+                        Err(err) => Err(err.into()),
+                    },
+                    TcpState::Fail(ref mut inner) => match inner.poll() {
+                        Err(err) => Err(err),
+                        _ => unreachable!(),
+                    },
+                }
             },
-            State::Fail(ref mut inner) => match inner.poll() {
-                Err(err) => Err(err),
-                _ => unreachable!(),
+            Self::Tls(state) => {
+                match state {
+                    TlsState::Wait(ref mut inner) => match inner.poll() {
+                        Ok(Async::Ready(tls)) => Ok(Async::Ready(tls.into())),
+                        Ok(Async::NotReady) => Ok(Async::NotReady),
+                        Err(err) => Err(err),
+                    },
+                    TlsState::Fail(ref mut inner) => match inner.poll() {
+                        Err(err) => Err(err),
+                        _ => unreachable!(),
+                    },
```

Comment:
probably it'd be more elegant to make State generic over implementors of `Poll<_>`

Reply:
It makes sense.